### PR TITLE
Migrate registry webhook to artifact event notifier

### PR DESCRIPTION
**Pull request by stonezdj**
Refactor event model
Move src/pkg/notification/model/const.go to src/pkg/notifier/model/const.go
Add auditlog handler to log project event, repo event, artifact event and tag event.
Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/pkg/notifier/handler/notification/access_log_hander.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"github.com/goharbor/harbor/src/pkg/notification/model"
+	"time"
+)
+
+type AccessLogHandler struct {
+}
+
+func (al *AccessLogHandler) Handle(value interface{}) error {
+
+	log.Debug("Start to handle access log ")
```
Comment (steven-zou) [inline src/pkg/notifier/handler/notification/access_log_hander.go]:
```suggestion	
	can we add logging middleware to log all the requests with a consistent format?
```

File: src/pkg/notifier/handler/notification/access_log_hander.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+		updatePullCount(imgEvent.RepoName)
+		updatePullTime(imgEvent.Project.ProjectID,
+			imgEvent.RepoName, imgEvent.Resource[0].Tag)
+
+	} else if imgEvent.EventType == model.EventTypeDeleteImage {
+		operationType = "delete"
+	}
+
+	if err := dao.AddAccessLog(models.AccessLog{
```
Comment (steven-zou) [inline src/pkg/notifier/handler/notification/access_log_hander.go]:
```suggestion
	if err := dao.AddAccessLog(models.AccessLog{
	
	no change to migrate such operations to the new programming model?
```

File: src/pkg/notifier/handler/notification/access_log_hander.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func updatePullCount(repository string) {
+	log.Debugf("Increase the repository %s pull count.", repository)
+	// if err := dao.IncreasePullCount(repository); err != nil {
+	//	log.Errorf("Error happens when increasing pull count: %v", repository)
+	// }
+}
+
+func updatePullTime(pID int64, repository, tag string) {
```
Comment (steven-zou) [inline src/pkg/notifier/handler/notification/access_log_hander.go]:
```suggestion
func updatePullTime(pID int64, repository, tag string) {

seems all these `updates` should be covered by the artifact ctl, why we still need to defined utility functions to handle such cases?
```

File: src/api/artifact/controller.go [hunk 4]
```diff
@@ -476,10 +481,61 @@ func (c *controller) ListTags(ctx context.Context, query *q.Query, option *TagOp
 func (c *controller) DeleteTag(ctx context.Context, tagID int64) error {
 	// Immutable checking is covered in middleware
 	// TODO check signature
-	// TODO fire delete tag event
+	c.sendTagDeleteEvent(ctx, tagID)
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: We support deleting tag and artifact in Harbor 2.0: 
- all tags attached to the artifact will be deleted when the artifact is deleted
- a tag can be deleted without deleting the artifact
That means deleting tag and deleting artifact are two separated event

File: src/api/artifact/controller.go [hunk 5]
```diff
@@ -476,10 +481,61 @@ func (c *controller) ListTags(ctx context.Context, query *q.Query, option *TagOp
...
+	}
+	imgEventMetadata := &event.ImageDelMetaData{
+		Project:  pro,
+		Tags:     []string{tag.Name},
+		Digests:  map[string]string{tag.Name: art.Digest},
+		RepoName: rep.Name,
+		OccurAt:  time.Now(),
+		Operator: username,
+	}
+	evt := &event.Event{}
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: No topic set?

File: src/server/registry/manifest.go [hunk 6]
```diff
@@ -109,5 +126,74 @@ func putManifest(w http.ResponseWriter, req *http.Request) {
 		log.Errorf("failed to flush: %v", err)
 	}
 
-	// TODO fire event, add access log in the event handler
+	sendImageEvent(req.Context(), tags, dgt, PushOperation)
```
Comment (wy65701436) [inline src/server/registry/manifest.go]: move it to artifact controller?

File: src/api/artifact/controller.go [hunk 7]
```diff
@@ -127,6 +129,7 @@ type controller struct {
 	abstractor   abstractor.Abstractor
 	immutableMtr match.ImmutableTagMatcher
 	regCli       registry.Client
+	proMgr       project.Manager
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: Seems this is useless as no reference

File: src/pkg/notifier/handler/auditlog/handler.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+type AuditResolver interface {
+	ResolveToAuditLog() (*am.AuditLog, error)
+}
+
+var AuditHandler = Handler{AuditLogMgr: audit.Mgr}
+
+func (h *Handler) Handle(value interface{}) error {
+	ctx := orm.NewContext(nil, beegoorm.NewOrm())
```
Comment (ywk253100) [inline src/pkg/notifier/handler/auditlog/handler.go]: The `Handle` function accepts the parameter `context` with the `ormer` and `transaction` populated? The `context` should be the same one within the request but populated new `transaction` as some event handling is a asynchronous process

File: src/pkg/notifier/handler/auditlog/handler.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+		if err != nil {
+			log.Errorf("failed to handler event %v", err)
+			return err
+		}
+		auditLog = al
+	default:
+		log.Errorf("Can not handler this event type! %#v", v)
+	}
+	if auditLog != nil {
+		h.AuditLogMgr.Create(ctx, auditLog)
```
Comment (ywk253100) [inline src/pkg/notifier/handler/auditlog/handler.go]: Handle error here

File: src/pkg/notifier/model/event.go [hunk 10]
```diff
@@ -59,6 +64,132 @@ type HookEvent struct {
...
+}
+
+func (p *ProjectEvent) ResolveToAuditLog() (*model.AuditLog, error) {
+	auditLog := &model.AuditLog{
+		ProjectID:    p.Project.ProjectID,
+		OpTime:       p.OccurAt,
+		Operation:    p.Operation,
+		Username:     p.Operator,
+		ResourceType: "project",
+		Resource: fmt.Sprintf("/api/project/%v/repository/%v/tag/%v",
```
Comment (ywk253100) [inline src/pkg/notifier/model/event.go]: Why it is an URL?

File: src/server/registry/manifest.go (hunk 6 above)
Comment (stonezdj) [reply src/server/registry/manifest.go]: remove it

**Pull request merged**